File: javascript/src/transform/snapshot-diff.js
import { readFile, writeFile, mkdir } from "node:fs/promises";
import { join } from "node:path";

/**
 * Parse a single CSV line, respecting double-quoted fields that may
 * contain commas or escaped quotes ("").
 *
 * @param {string} line - A single CSV row as a string.
 * @returns {string[]} Array of field values.
 */
function parseCsvLine(line) {
  const fields = [];
  let current = "";
  let inQuotes = false;
  let i = 0;

  while (i < line.length) {
    const ch = line[i];

    if (inQuotes) {
      if (ch === '"') {
        if (i + 1 < line.length && line[i + 1] === '"') {
          current += '"';
          i += 2;
          continue;
        }
        inQuotes = false;
        i++;
        continue;
      }
      current += ch;
      i++;
    } else {
      if (ch === '"') {
        inQuotes = true;
        i++;
      } else if (ch === ",") {
        fields.push(current);
        current = "";
        i++;
      } else {
        current += ch;
        i++;
      }
    }
  }

  fields.push(current);
  return fields;
}

/**
 * Serialize a row of field values into a CSV line, quoting fields that
 * contain commas, double quotes, or newlines.
 *
 * @param {string[]} row - Field values.
 * @returns {string} A single CSV-formatted line.
 */
function rowToCsvLine(row) {
  return row
    .map((field) => {
      if (
        field.includes(",") ||
        field.includes('"') ||
        field.includes("\n") ||
        field.includes("\r")
      ) {
        return '"' + field.replace(/"/g, '""') + '"';
      }
      return field;
    })
    .join(",");
}

/**
 * Parse a CSV string into headers and row objects.
 * Uses a simple parser that handles quoted fields with commas and escaped
 * quotes (""). Strips whitespace from headers and values. Skips blank lines.
 *
 * @param {string} content - Raw CSV string.
 * @returns {{ headers: string[], rows: Object[] }} Headers and array of row objects.
 */
export function parseCsv(content) {
  const lines = content
    .split(/\r?\n/)
    .filter((line) => line.trim().length > 0);

  if (lines.length === 0) {
    return { headers: [], rows: [] };
  }

  const headers = parseCsvLine(lines[0]).map((h) => h.trim());
  const rows = [];

  for (let i = 1; i < lines.length; i++) {
    const values = parseCsvLine(lines[i]).map((v) => v.trim());
    const row = {};
    for (let j = 0; j < headers.length; j++) {
      row[headers[j]] = values[j] ?? "";
    }
    rows.push(row);
  }

  return { headers, rows };
}

/**
 * Read a CSV file and return headers and row objects.
 *
 * @param {string} filePath - Path to the CSV file.
 * @returns {Promise<{ headers: string[], rows: Object[] }>} Parsed CSV data.
 */
export async function readSnapshot(filePath) {
  const content = await readFile(filePath, "utf-8");
  return parseCsv(content);
}

/**
 * Index rows by composite primary key.
 *
 * Builds a Map from a key string (column values joined by \0) to the row
 * object. Throws if a key column is missing from the data or if a duplicate
 * key is found.
 *
 * @param {Object[]} rows - Array of row objects.
 * @param {string[]} keyColumns - Column names forming the primary key.
 * @returns {Map<string, Object>} Map from key string to row object.
 * @throws {Error} If a key column is missing from the data.
 * @throws {Error} If a duplicate key is found.
 */
export function indexByKey(rows, keyColumns) {
  if (rows.length > 0) {
    const firstRowKeys = Object.keys(rows[0]);
    for (const col of keyColumns) {
      if (!firstRowKeys.includes(col)) {
        throw new Error(`Key column "${col}" not found in data`);
      }
    }
  }

  const map = new Map();

  for (const row of rows) {
    const keyStr = keyColumns.map((col) => row[col]).join("\0");
    if (map.has(keyStr)) {
      throw new Error(`Duplicate key found: ${keyStr}`);
    }
    map.set(keyStr, row);
  }

  return map;
}

/**
 * Compare two row objects and return a list of column changes.
 * Excludes key columns and ignored columns from comparison.
 *
 * @param {Object} oldRow - Row object from the old snapshot.
 * @param {Object} newRow - Row object from the new snapshot.
 * @param {string[]} keyColumns - Column names forming the primary key.
 * @param {string[]} [ignoreColumns=[]] - Column names to exclude from comparison.
 * @returns {{ column: string, old_value: string, new_value: string }[]} List of changes.
 */
export function compareRows(oldRow, newRow, keyColumns, ignoreColumns = []) {
  const skip = new Set([...keyColumns, ...ignoreColumns]);
  const changes = [];

  const allColumns = new Set([...Object.keys(oldRow), ...Object.keys(newRow)]);

  for (const col of allColumns) {
    if (skip.has(col)) continue;
    const oldVal = oldRow[col] ?? "";
    const newVal = newRow[col] ?? "";
    if (oldVal !== newVal) {
      changes.push({ column: col, old_value: oldVal, new_value: newVal });
    }
  }

  return changes;
}

/**
 * Build a key object from a row and key columns.
 *
 * @param {Object} row - A row object.
 * @param {string[]} keyColumns - Column names forming the primary key.
 * @returns {Object} Object with only the key columns.
 */
function buildKeyObject(row, keyColumns) {
  const key = {};
  for (const col of keyColumns) {
    key[col] = row[col];
  }
  return key;
}

/**
 * Compare two CSV snapshots and produce a diff summary.
 *
 * Reads both files, indexes by key columns, then classifies every row as
 * an insert, update, delete, or unchanged.
 *
 * @param {string} oldPath - Path to old snapshot CSV.
 * @param {string} newPath - Path to new snapshot CSV.
 * @param {string[]} keyColumns - Column names forming the primary key.
 * @param {string[]} [ignoreColumns=[]] - Column names to exclude from comparison.
 * @returns {Promise<{
 *   old_row_count: number,
 *   new_row_count: number,
 *   inserts: number,
 *   updates: number,
 *   deletes: number,
 *   unchanged: number,
 *   changes: Array<{
 *     change_type: "insert"|"update"|"delete",
 *     key: Object,
 *     row: Object,
 *     changed_columns: Array<{column: string, old_value: string, new_value: string}>
 *   }>
 * }>}
 */
export async function compareSnapshots(
  oldPath,
  newPath,
  keyColumns,
  ignoreColumns = [],
) {
  const oldSnapshot = await readSnapshot(oldPath);
  const newSnapshot = await readSnapshot(newPath);

  const oldIndex = indexByKey(oldSnapshot.rows, keyColumns);
  const newIndex = indexByKey(newSnapshot.rows, keyColumns);

  const changes = [];
  let inserts = 0;
  let updates = 0;
  let deletes = 0;
  let unchanged = 0;

  // Iterate new index: keys not in old are inserts, keys in both need comparison
  for (const [keyStr, newRow] of newIndex) {
    const keyObj = buildKeyObject(newRow, keyColumns);

    if (!oldIndex.has(keyStr)) {
      inserts++;
      changes.push({
        change_type: "insert",
        key: keyObj,
        row: newRow,
        changed_columns: [],
      });
    } else {
      const oldRow = oldIndex.get(keyStr);
      const diffs = compareRows(oldRow, newRow, keyColumns, ignoreColumns);
      if (diffs.length > 0) {
        updates++;
        changes.push({
          change_type: "update",
          key: keyObj,
          row: newRow,
          changed_columns: diffs,
        });
      } else {
        unchanged++;
      }
    }
  }

  // Iterate old index: keys not in new are deletes
  for (const [keyStr, oldRow] of oldIndex) {
    if (!newIndex.has(keyStr)) {
      deletes++;
      const keyObj = buildKeyObject(oldRow, keyColumns);
      changes.push({
        change_type: "delete",
        key: keyObj,
        row: oldRow,
        changed_columns: [],
      });
    }
  }

  // Sort changes by key string for determinism
  changes.sort((a, b) => {
    const aKey = keyColumns.map((col) => a.key[col]).join("\0");
    const bKey = keyColumns.map((col) => b.key[col]).join("\0");
    return aKey < bKey ? -1 : aKey > bKey ? 1 : 0;
  });

  return {
    old_row_count: oldSnapshot.rows.length,
    new_row_count: newSnapshot.rows.length,
    inserts,
    updates,
    deletes,
    unchanged,
    changes,
  };
}

/**
 * Write diff output files.
 *
 * Creates CSV files for inserts, updates, and deletes (only when non-empty),
 * plus a summary.json with the full diff summary. The updates CSV includes an
 * extra `changed_columns` column with a JSON representation of the changes.
 *
 * @param {Object} summary - DiffSummary from compareSnapshots.
 * @param {string} outputDir - Directory to write output files into.
 * @param {string[]} headers - Column headers for CSV output.
 * @returns {Promise<Object>} Map of file type to path.
 */
export async function writeDiffFiles(summary, outputDir, headers) {
  await mkdir(outputDir, { recursive: true });

  const files = {};

  const insertRows = summary.changes.filter(
    (c) => c.change_type === "insert",
  );
  const updateRows = summary.changes.filter(
    (c) => c.change_type === "update",
  );
  const deleteRows = summary.changes.filter(
    (c) => c.change_type === "delete",
  );

  if (insertRows.length > 0) {
    const path = join(outputDir, "inserts.csv");
    const lines = [
      rowToCsvLine(headers),
      ...insertRows.map((c) =>
        rowToCsvLine(headers.map((h) => c.row[h] ?? "")),
      ),
    ];
    await writeFile(path, lines.join("\n") + "\n", "utf-8");
    files.inserts = path;
  }

  if (updateRows.length > 0) {
    const updateHeaders = [...headers, "changed_columns"];
    const path = join(outputDir, "updates.csv");
    const lines = [
      rowToCsvLine(updateHeaders),
      ...updateRows.map((c) => {
        const values = headers.map((h) => c.row[h] ?? "");
        values.push(JSON.stringify(c.changed_columns));
        return rowToCsvLine(values);
      }),
    ];
    await writeFile(path, lines.join("\n") + "\n", "utf-8");
    files.updates = path;
  }

  if (deleteRows.length > 0) {
    const path = join(outputDir, "deletes.csv");
    const lines = [
      rowToCsvLine(headers),
      ...deleteRows.map((c) =>
        rowToCsvLine(headers.map((h) => c.row[h] ?? "")),
      ),
    ];
    await writeFile(path, lines.join("\n") + "\n", "utf-8");
    files.deletes = path;
  }

  const summaryPath = join(outputDir, "summary.json");
  await writeFile(summaryPath, JSON.stringify(summary, null, 2), "utf-8");
  files.summary = summaryPath;

  return files;
}

/**
 * Format a diff summary as a human-readable string.
 *
 * @param {Object} summary - DiffSummary from compareSnapshots.
 * @returns {string} Formatted summary text.
 */
export function formatSummary(summary) {
  return [
    "=== Snapshot Diff Summary ===",
    `Old snapshot: ${summary.old_row_count} rows`,
    `New snapshot: ${summary.new_row_count} rows`,
    `Inserts: ${summary.inserts} | Updates: ${summary.updates} | Deletes: ${summary.deletes} | Unchanged: ${summary.unchanged}`,
  ].join("\n");
}

File: python/src/data_platform_lab/transform/snapshot_diff.py
"""CDC snapshot comparison tool.

Compares two CSV snapshots (old vs new) and produces a detailed diff
identifying inserted, updated, deleted, and unchanged rows.  Useful for
change-data-capture pipelines that rely on periodic full-table exports.
"""

from __future__ import annotations

import csv
import json
import logging
from dataclasses import asdict, dataclass, field
from pathlib import Path

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------


@dataclass
class ColumnChange:
    """A single column-level change within an updated row."""

    column: str
    old_value: str
    new_value: str


@dataclass
class RowChange:
    """A changed row with its change type and details."""

    change_type: str  # "insert", "update", "delete"
    key: dict[str, str]  # primary key columns -> values
    row: dict[str, str]  # full row data (new for insert/update, old for delete)
    changed_columns: list[ColumnChange] = field(default_factory=list)


@dataclass
class DiffSummary:
    """Summary of a snapshot comparison."""

    old_row_count: int
    new_row_count: int
    inserts: int
    updates: int
    deletes: int
    unchanged: int
    changes: list[RowChange] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Reading
# ---------------------------------------------------------------------------


def read_snapshot(path: Path) -> tuple[list[str], list[dict[str, str]]]:
    """Read a CSV snapshot, return (headers, list of row dicts).

    Uses the stdlib csv module.  Strip whitespace from headers and values.
    """
    with path.open("r", newline="", encoding="utf-8") as fh:
        reader = csv.DictReader(fh)
        raw_fieldnames: list[str] = list(reader.fieldnames or [])
        headers = [h.strip() for h in raw_fieldnames]

        rows: list[dict[str, str]] = []
        for raw_row in reader:
            row = {k.strip(): v.strip() for k, v in raw_row.items()}
            rows.append(row)

    logger.info("Read %d rows from %s", len(rows), path)
    return headers, rows


# ---------------------------------------------------------------------------
# Indexing
# ---------------------------------------------------------------------------


def index_by_key(
    rows: list[dict[str, str]],
    key_columns: list[str],
) -> dict[tuple[str, ...], dict[str, str]]:
    """Index a list of row dicts by composite primary key.

    Returns a dict mapping key tuples to row dicts.
    Raises ``ValueError`` if a duplicate key is found.
    Raises ``KeyError`` if a key column is missing from the data.
    """
    if rows:
        first_row = rows[0]
        for col in key_columns:
            if col not in first_row:
                raise KeyError(
                    f"Key column '{col}' not found in data. "
                    f"Available columns: {sorted(first_row.keys())}"
                )

    index: dict[tuple[str, ...], dict[str, str]] = {}
    for row in rows:
        key = tuple(row[col] for col in key_columns)
        if key in index:
            raise ValueError(
                f"Duplicate key {dict(zip(key_columns, key, strict=False))} found in data"
            )
        index[key] = row

    return index


# ---------------------------------------------------------------------------
# Comparison helpers
# ---------------------------------------------------------------------------


def compare_rows(
    old_row: dict[str, str],
    new_row: dict[str, str],
    key_columns: list[str],
    ignore_columns: list[str] | None = None,
) -> list[ColumnChange]:
    """Compare two rows and return a list of :class:`ColumnChange` objects.

    Excludes key columns and ignored columns from comparison.
    """
    skip = set(key_columns)
    if ignore_columns:
        skip.update(ignore_columns)

    changes: list[ColumnChange] = []
    for col in old_row:
        if col in skip:
            continue
        old_val = old_row[col]
        new_val = new_row.get(col, "")
        if old_val != new_val:
            changes.append(ColumnChange(column=col, old_value=old_val, new_value=new_val))

    return changes


# ---------------------------------------------------------------------------
# Main comparison
# ---------------------------------------------------------------------------


def compare_snapshots(
    old_path: Path,
    new_path: Path,
    key_columns: list[str],
    ignore_columns: list[str] | None = None,
) -> DiffSummary:
    """Compare two CSV snapshots and produce a :class:`DiffSummary`.

    Algorithm:
    1. Read both snapshots
    2. Index both by *key_columns*
    3. Keys in new but not old -> inserts
    4. Keys in old but not new -> deletes
    5. Keys in both -> compare columns; if any differ -> update, else -> unchanged
    6. Sort changes by key for determinism

    Raises ``ValueError`` if duplicate keys found in either snapshot.
    Raises ``KeyError`` if *key_columns* are missing from headers.
    """
    _old_headers, old_rows = read_snapshot(old_path)
    _new_headers, new_rows = read_snapshot(new_path)

    old_index = index_by_key(old_rows, key_columns)
    new_index = index_by_key(new_rows, key_columns)

    changes: list[RowChange] = []
    unchanged_count = 0

    # Inserts: keys in new but not in old
    for key, row in new_index.items():
        if key not in old_index:
            changes.append(
                RowChange(
                    change_type="insert",
                    key=dict(zip(key_columns, key, strict=False)),
                    row=row,
                )
            )

    # Deletes: keys in old but not in new
    for key, row in old_index.items():
        if key not in new_index:
            changes.append(
                RowChange(
                    change_type="delete",
                    key=dict(zip(key_columns, key, strict=False)),
                    row=row,
                )
            )

    # Updates / unchanged: keys in both
    for key in old_index:
        if key in new_index:
            col_changes = compare_rows(old_index[key], new_index[key], key_columns, ignore_columns)
            if col_changes:
                changes.append(
                    RowChange(
                        change_type="update",
                        key=dict(zip(key_columns, key, strict=False)),
                        row=new_index[key],
                        changed_columns=col_changes,
                    )
                )
            else:
                unchanged_count += 1

    # Sort for determinism
    changes.sort(key=lambda c: tuple(c.key.values()))

    insert_count = sum(1 for c in changes if c.change_type == "insert")
    update_count = sum(1 for c in changes if c.change_type == "update")
    delete_count = sum(1 for c in changes if c.change_type == "delete")

    summary = DiffSummary(
        old_row_count=len(old_rows),
        new_row_count=len(new_rows),
        inserts=insert_count,
        updates=update_count,
        deletes=delete_count,
        unchanged=unchanged_count,
        changes=changes,
    )

    logger.info(
        "Snapshot diff complete: %d inserts, %d updates, %d deletes, %d unchanged",
        insert_count,
        update_count,
        delete_count,
        unchanged_count,
    )

    return summary


# ---------------------------------------------------------------------------
# Output
# ---------------------------------------------------------------------------


def write_diff_files(
    summary: DiffSummary,
    output_dir: Path,
) -> dict[str, Path]:
    """Write inserts.csv, updates.csv, deletes.csv, and summary.json to *output_dir*.

    For ``updates.csv``, include all columns plus a ``changed_columns`` column
    listing the names of columns that changed (comma-separated).

    Returns a dict mapping file type to path.  Only writes files that have data.
    """
    output_dir.mkdir(parents=True, exist_ok=True)
    result: dict[str, Path] = {}

    inserts = [c for c in summary.changes if c.change_type == "insert"]
    updates = [c for c in summary.changes if c.change_type == "update"]
    deletes = [c for c in summary.changes if c.change_type == "delete"]

    # --- inserts.csv ---
    if inserts:
        path = output_dir / "inserts.csv"
        fieldnames = list(inserts[0].row.keys())
        with path.open("w", newline="", encoding="utf-8") as fh:
            writer = csv.DictWriter(fh, fieldnames=fieldnames)
            writer.writeheader()
            for change in inserts:
                writer.writerow(change.row)
        result["inserts"] = path

    # --- updates.csv ---
    if updates:
        path = output_dir / "updates.csv"
        fieldnames = [*list(updates[0].row.keys()), "changed_columns"]
        with path.open("w", newline="", encoding="utf-8") as fh:
            writer = csv.DictWriter(fh, fieldnames=fieldnames)
            writer.writeheader()
            for change in updates:
                row = dict(change.row)
                row["changed_columns"] = ",".join(cc.column for cc in change.changed_columns)
                writer.writerow(row)
        result["updates"] = path

    # --- deletes.csv ---
    if deletes:
        path = output_dir / "deletes.csv"
        fieldnames = list(deletes[0].row.keys())
        with path.open("w", newline="", encoding="utf-8") as fh:
            writer = csv.DictWriter(fh, fieldnames=fieldnames)
            writer.writeheader()
            for change in deletes:
                writer.writerow(change.row)
        result["deletes"] = path

    # --- summary.json ---
    summary_path = output_dir / "summary.json"
    summary_dict = asdict(summary)
    with summary_path.open("w", encoding="utf-8") as fh:
        json.dump(summary_dict, fh, indent=2)
    result["summary"] = summary_path

    logger.info("Wrote diff files to %s", output_dir)
    return result


# ---------------------------------------------------------------------------
# Formatting
# ---------------------------------------------------------------------------


def format_summary(summary: DiffSummary) -> str:
    """Return a human-readable summary string."""
    lines = [
        "Snapshot Diff Summary",
        "=" * 40,
        f"Old snapshot rows : {summary.old_row_count}",
        f"New snapshot rows : {summary.new_row_count}",
        "-" * 40,
        f"Inserts           : {summary.inserts}",
        f"Updates           : {summary.updates}",
        f"Deletes           : {summary.deletes}",
        f"Unchanged         : {summary.unchanged}",
        "-" * 40,
    ]

    if summary.changes:
        lines.append("Details:")
        for change in summary.changes:
            key_str = ", ".join(f"{k}={v}" for k, v in change.key.items())
            if change.change_type == "update":
                col_details = "; ".join(
                    f"{cc.column}: {cc.old_value!r} -> {cc.new_value!r}"
                    for cc in change.changed_columns
                )
                lines.append(f"  [{change.change_type.upper()}] {key_str} ({col_details})")
            else:
                lines.append(f"  [{change.change_type.upper()}] {key_str}")

    lines.append("=" * 40)
    return "\n".join(lines)
